pyomo/util/config_domains.py

```diff
@@ -0,0 +1,55 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.common.collections import ComponentSet
+
+
+class ComponentDataList:
+    """ComponentDataList(ctype)
+    Domain validation class that accepts singleton or iterable arguments and
+    compiles them into a ComponentSet, verifying that they are all ComponentDatas
+    of type 'ctype.'
+
+    Parameters
+    ----------
+        ctype: The component type of the list
+
+    Raises
+    ------
+        ValueError if all of the arguments are not of type 'ctype'
+    """
+
+    def __init__(self, ctype):
+        self._ctype = ctype
+
+    def __call__(self, x):
+        if hasattr(x, 'ctype') and x.ctype is self._ctype:
+            if not x.is_indexed():
+                return ComponentSet([x])
+            ans = ComponentSet()
+            for j in x.index_set():
+                ans.add(x[j])
+            return ans
+        elif hasattr(x, '__iter__'):
+            ans = ComponentSet()
+            for i in x:
+                ans.update(self(i))
+            return ans
+        else:
+            _ctype_name = str(self._ctype)
+            raise ValueError(
+                f"Expected {_ctype_name} or iterable of "
+                f"{_ctype_name}s.\n\tReceived {type(x)}"
+            )
```

Comment:
There are some issues with this implementation:
1. it doesn't actually return a list. ;)
2. it doesn't actually validate the ctype for list-like inputs

Maybe the following would be a more robust alternative:
```python
def __init__(self, ctype):
    if isinstance(ctype, Sequence):
        self._ctypes = set(ctype)
    else:
        self._ctypes = set([ctype])

def __call__(self, x):
    return list(self._process(x))

def _process(self, x):
    if hasattr(x, 'ctype'):
        if x.ctype not in self._ctypes:
            raise ValueError(" .... ")
        if x.is_indexed():
            yield from x.values()
        else:
            yield x
    elif isinstance(x, Sequence):
        for y in x:
            yield from self._process(y)
    else:
        raise ValueError(" ... ")

def domain_name(self):
    _names = ', '.join(ct.__name__ for ct in self._ctypes)
    if len(self._ctypes) > 1:
        _names = '[' + _names + ']'
    return f"(ComponentDataList({_names}")

Reply:
I actually really want it to both return and accept ComponentSets, but just took your suggestions otherwise (and renamed it accordingly)